[team-2][만사, Downey] RxSwift & AlamoFire를 사용한 Network 및 View 구성.

dahun-lee-daji opened the pull request:
# 작업 내용
1. Network는 RxSwift와 AlamoFire를 사용하여 구현했습니다. 
2. MockAPI을 이용하여 Network작동을 확인했습니다.  
3. UI를 작성할 때 Xib로 파일을 분리하여 재 사용성 및 가독성을 확보하려 노력했습니다.
4. RxSwift를 사용하며 MVVM의 클린아키텍쳐를 지키도록 노력했습니다.  

# 진행사항 체크리스트
### Story -1 참가할 수 있는 게임 목록을 볼 수 있다.
- [x] 1-1. 게임 목록은 CollectionView로 구성한다.  
- [x] 1-2. 참가 가능한 게임 목록을 표시한다. 이 때, 게임의 인덱스, 원정팀 (좌측) 홈팀 (우측)을 표기한다.  
- [ ] 1-3. 홈 팀과 원정팀의 선택은 Action Sheet를 사용한다.  
- [ ] 1-4. 다른 사람이 선택한 경우, 참가할 수 없고, 다른 게임을 선택하도록 안내한다. ( Network ) 

### Story - 2  공격,수비 이닝을 진행 할 수 있다.
- [x] 공통-1. Tabbar를 embed하여 게임 화면과 상세 스코어 페이지를 이동 할 수 있다.  
- [ ] 공통-2. 제목은 __원정팀이름__ , 점수 vs 점수, __홈팀이름__ 형태를 표시한다.   __Bold : 미구현__
- [ ] 공통-3. 주자의 이동은, 애니메이션으로 색이 나타나고, 흐려지고를 통해 보여준다.
- [x] 공통-4. 사용자는 SBO를 색이 채워진 동그라미의 수로 알 수 있다.
- [x] 공통-5. 투수 이름과, 투구수를 표시한다.
- [x] 공통-6. 타석에 있는 타자 이름과 타석, 안타수를 표시한다.
- [ ] 공통-7. 게임이 진행 될 때 마다, 리스트 항목을 업데이트 한다.
- [ ] 공통-8. Scores 탭바를 누르면 타자별 스코어를 확인 할 수 있다.
- [x] 공통-9. 게임 이닝 수 와 선택한 팀의 공 수 상황을 표기한다.
- [ ] 공-1. 공격은 자동으로 3초마다 진행한다. ( network )
- [ ] 공-2. 공격 일 때, 타자의 우측 끝에 체크, 수비 일 때, 투수의 우측 끝에 체크를 통해 공수를 알 수 있게 한다.
- [ ] 수-1. PITCH 버튼을 눌러 게임을 진행한다. ( network )

### Story - 4 상세 스코어 화면을 볼 수 있다
- [ ] 4-1. 이닝 별 점수를 표기하고, 현재 공격 팀을 점수판에 표기한다.  
- [ ] 4-2. segmented controller를 통해 선택한 팀의 명단을 표기한다.  
- [ ] 4-3. 팀의 명단에는 타석 안타 아웃 평균을 표기한다.   
- [ ] 4-4. 명단의 끝에 Total을 표기한다.
- [ ] 4-5. 현재 공격, 수비중인 타자와 투수를 highlight한다.  
- [x] 4-6. Play Tabbar버튼을 눌러 게임 화면으로 돌아간다.

# 학습 키워드
1. Network : RxSwift + AlamoFire
2. MVVM
3. Clean Architecture
4. RxSwift

# 고민 & 질문 사항
1. RxSwift를 처음 사용하며, 기본적인 사용법은 익혔지만, 좋은 구조와 가독성 좋은 코드를 연습해야 할 것 같다.  ( ViewModel에서 Input - Output 구조를 시도해야 함. -> Test 유도, 재 사용성 매우 좋음)
2. 이전 프로젝트에서 Combine을 사용하여, URLSession.dataTaskPublish를 사용 할 때와  현재 AlamoFire를 사용 할 때, 장점을 느 끼기 어렵습니다. AlamoFire를 옳게 사용하지 못 한 것인지 의아합니다.

# 반성할 점
1. 충분한 조사를 진행하지 않아, AlamoFire를 선택 한 후 RxAlamoFire의 존재를 알게 되었습니다. RxAlamoFire를 사용함이 옳았을지 후회됩니다.
2. UI 구성 시, 생각보다 중복되는 코드가 발생함. Register를 Protocol로 묶어서 처리 할 수 있었는데, 시행하지 않음.

Commit:
docs: Initial README

dahun-lee-daji committed:
Feat: init Proj. with cocoapods (#none)

cocoa pods : RxSwift, RxCocoa

Commit:
gitignore추가

dahun-lee-daji committed:
Feat: Add NetworkService class (#10)

just only file and folder; no codes.

dahun-lee-daji committed:
feat: Add Alamofire (#10)

with cocoapods

Commit:
feat: 게임목록 화면 구현 (#9)

Commit:
feat: tabbar 화면 구현 및 스코어 nib 파일 적용 (#13)

dahun-lee-daji committed:
Feat: Add get method to NetworkService (#10)

dahun-lee-daji committed:
Feat: Add Entities, APIPath cases (#10)

Modify NetworkService.get method's signiture

ITzombietux committed:
Merge pull request #27 from ghojeong/feat/iOS/10/NetworkService

Feat/ios/10/network service

Commit:
feat: socore tab UI 틀 구현 (#28)

Commit:
feat: VC 추가 (#28)

Commit:
feat: play tab UI 틀 구현 (#21)

Commit:
Merge branch 'ios' of https://github.com/ghojeong/baseball into ios

Commit:
fix: Merge branch ios (#29)

Commit:
feat:RxSwift를 활용한 데이터 바인딩 (#29)

Commit:
feat:RxSwift를 활용한 데이터 바인딩 (#39)

Commit:
fix:Entity 수정 (#39)

Commit:
fix:ScoreUseCase추가 및 다른 UseCase  수정 (#39)

Commit:
feat:Score tab VC 추가 (#41)

Commit:
fix: Score tab UI수정 (#41)

dahun-lee-daji committed:
Merge pull request #42 from ghojeong/feat/ios/29/gameListArchitecture

[iOS] 게임 목록 RxSwift를 사용한 architecture

Commit:
fix: Play tab UI수정 (#39)

dahun-lee-daji committed:
Merge pull request #43 from ghojeong/feat/ios/39/playTabArchitecture

[iOS] play tab architecture

GangWoon commented:
어떤 이유인지는 모르겠지만 지금 디프가 작동을 안하네요...
커밋을 따라가면서 코맨트를 남기는 식으로 진행하도록 하겠습니다.

GangWoon commented on ios/baseball/baseball/Service/Network/EndPoint.swift:
```diff
@@ -0,0 +1,31 @@
+struct EndPoint {
+    
+    enum APIPath: String {
+        case test = "test"
```
enum 타입에 rawValue로 직접 접근하는 방식보단 CustomStringConvertible을 채택해보시는 건 어떨까요?

GangWoon commented on ios/baseball/baseball/Presentor/UI/Controller/GameListViewController.swift:
```diff
@@ -0,0 +1,46 @@
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        self.collectionView.dataSource = self
```
스토리보드에선 구현하실 순 없었을까요?

GangWoon commented on ios/baseball/baseball/Service/Network/EndPoint.swift:
```diff
@@ -6,26 +6,42 @@
         component.path = "\(path.rawValue)"
         
-        return component.url
+        return component.url!
```
Optional은 분기를 처리할 수 있는 좋은 기회가 될 수 있습니다.
되도록 강제 언래핑을 사용하시는 걸 피해주셨으면 좋겠습니다.

GangWoon commented on ios/baseball/baseball/Service/Network/EndPoint.swift:
```diff
@@ -6,26 +6,42 @@
+            request = try URLRequest.init(url: self.url, method: self.method, headers: HTTPHeaders.init(arrayLiteral: HTTPHeader.accept("application/json")))
+            
+        } catch  {
+            assertionFailure("EndPoint.asURLRequest")
```
request 생성을 실패했는데 앱을 꺼버리는건 사용자가 사용하기에 불편하지 않을까요?

dahun-lee-daji replied:
사용성을 고려하는 것을 매우 필요한 일이라고 생각합니다.
그런데,,,, 프로젝트를 진행하면서 network연결이 어렵다고 modal창 띄워주는 그 기능 하나 구현하는게 너무 부담스럽네요 ㅜㅜ
사용성에 대한 것은 항상 인지하도록 하겠습니다!

GangWoon commented on ios/baseball/baseball/Service/Network/NetworkService.swift:
```diff
@@ -30,6 +31,7 @@ class NetworkService {
                 switch response.result {
                 case .success(let data):
                     do {
+                        print(T.self)
```
불필요한 코드는 제거해주셨으면 좋겠습니다.
프린트문을 사용해서 확인하는 방법 말고도 해당 데이터를 확인하는 방법은 없을까요?

dahun-lee-daji replied:
breakPoint를 사용해서 확인 할 수 있다는 것을 알 고 있지만, 선뜻 손이 가지 않는 것 같습니다 ㅜㅜ

불필요한 코드는 제거 완료했습니다! 😃

GangWoon commented on ios/baseball/baseball/Presentor/UI/View/Play/MatchUpView.swift:
```diff
@@ -0,0 +1,22 @@
+    }
+    
+    func commonInit() {
+        let view = Bundle.main.loadNibNamed("MatchUpView", owner: self, options: nil)?.first as! UIView
```
정말로 사소해보이실 수도 있지만 정말 위험한 접근입니다. 
강제 언래핑은요

ITzombietux replied:
네! 옵셔널 바인딩 처리를 했습니다.!

GangWoon commented on ios/baseball/baseball/Presentor/UI/View/Score/InningNumberStack.swift:
```diff
@@ -0,0 +1,41 @@
+    }
+    
+    private func addInningNumberLabels() {
+        let inningNumberRange = 1...11
```
숫자를 의미있는 숫자로 표현해보시는 건 어떨까요?

GangWoon commented on ios/baseball/baseball/Presentor/UI/View/Score/TeamInningScoreStack.swift:
```diff
@@ -0,0 +1,29 @@
+    
+    private func addInningViewLabels() {
+        let max = 12
+        for _ in 0..<max {
```
위에서 사용하신 방법과 다른 형식이네요.
for문과 forEach 둘다 사용하셨는데 각각 사용하신 이유가 있을까요?

ITzombietux replied:
아….! 서치하며 예시 코드를 썼던게 for _ in 이었고 제가 즐겨쓰는게 forEach다 보니 둘 다 썼던거 같아요. 이런 사소한 것도 염두헤두고 개발을 하겠습니다.

GangWoon commented on ios/baseball/baseball/Presentor/UI/View/Play/PitchButton.swift:
```diff
@@ -7,14 +7,19 @@
+    
+    private func configureUI() {
+        layer.cornerRadius = 6
+//        isHidden = true
```
사용하지 않는 코드는 읽는 사람에게 혼동을 유발할 수 있습니다.

GangWoon commented on ios/baseball/baseball/Presentor/UI/View/Play/SBOView.swift:
```diff
@@ -7,14 +7,52 @@
+    @IBOutlet weak var countStack: UIStackView!
+    
+    private var countViews: [CountView]!
+    var countColor: UIColor!
```
타입이 UIColor!일 필요가 있을까요?

GangWoon commented on ios/baseball/baseball/Domain/Entity/InningInfo.swift:
```diff
@@ -0,0 +1,22 @@
+    private(set) var userOffense: Bool
+    
+    func makeTopOrBottom() -> String {
+        return userTop ? "초" : "말"
```
메소드 이름과 내부 내용의 흐름이 일치하지 않습니다.
다시 표현해보시는 건 어떠실까요?

ITzombietux replied:
네 ! 그렇네요 ㅎㅎㅎ 좀 더 읽기 좋은 표현으로 바꾸겠습니다.

GangWoon commented on ios/baseball/baseball/Domain/Entity/SBO.swift:
```diff
@@ -2,13 +2,13 @@
-    private(set) var strikeCount: Int
-    private(set) var ballCount: Int
-    private(set) var outCount: Int
+    private(set) var strike: Int
```
변경하신 이유를 들을 수 있을까요?

ITzombietux replied:
이것도 SBOCount라고 Struct의 이름을 지정해야 할지 SBO라고 Struct의 이름을 지정하고 안에서 속성에 -Count를 붙일지 고민 중에 그냥 넘어가버린 것 같습니다. 좀 더 이해하기 좋게 수정하도록 할게요.!

GangWoon commented on ios/baseball/baseball/Domain/UseCase/PlayUseCase.swift:
```diff
@@ -0,0 +1,26 @@
+import RxCocoa
+
+protocol PlayUseCasePort {
+    func get(path: APIPath, id: String?) -> Observable<MatchInfo>
```
"MatchUseCasePort"과 유사해 보이는데 한가지 메소드로 묶을 순 없을까요?

GangWoon commented on ios/baseball/baseball/Domain/UseCase/PlayUseCase.swift:
```diff
@@ -0,0 +1,26 @@
+class PlayUseCase: PlayUseCasePort {
+    private var networkService: NetworkServiceable
+    
+    init(networkService: NetworkServiceable = NetworkService()) {
```
프로토콜이라고 -able을 꼭 붙여야 된다는 생각은 안하셔도 괜찮을 거 같습니다.

ITzombietux replied:
네. 보통 -able, -Type, -Port 느낌으로 많이 사용하길래 사용했는데, 적절한 네이밍을 다시 한번 생각해볼게요.

GangWoon commented on ios/baseball/baseball/Domain/UseCase/PlayViewModel.swift:
```diff
@@ -0,0 +1,48 @@
+class PlayViewModel {
+    private(set) var matchInfo = BehaviorSubject<MatchInfo>(value: MatchInfo())
+    private var playUseCase: PlayUseCasePort!
+    lazy private(set) var pitchInfo = matchInfo.map {
```
lazy로 생성하신 이유를 들을 수 있을까요?

ITzombietux replied:
BehaviorSubject로 만들어진 matchInfo의 업데이트된 데이터를 사용하기 위해 lazy로 만들었습니다. lazy로 생성하지 않고 만들 수 있는 방향도 존재했습니다만, 아직 RxSwift를 처음 사용하다보니 편한 방법으로 진행했던 것 같습니다. ! lazy를 사용하지 않는 방향으로도 수정을 시도해볼게요.

GangWoon commented on ios/baseball/baseball/Presentor/UI/Controller/PlayViewController.swift:
```diff
@@ -6,10 +6,81 @@
+    @IBOutlet weak var matchUpView: MatchUpView!
+    @IBOutlet weak var collectionView: UICollectionView!
+    
+    private var id: String = "U924AX"
```
id와 같이 애매모호한 이름 역시 읽는 사람에게
해당 코드가 어떤걸 의미하는지 많은 생각을 하게 만듭니다.

GangWoon commented on ios/baseball/baseball/Presentor/UI/View/Play/MatchBoardView.swift:
```diff
@@ -24,9 +24,13 @@ final class MatchBoardView: UIView {
+    }
+    
+    func configureInningInfo(inningInfo: InningInfo) {
+        inningInfoLabel.text = "\(inningInfo.inningCount)회\(inningInfo.makeTopOrBottom()) \(inningInfo.makeOffenseOrDefence())"
```
해당 작업을 viewModel로 옮기는건 어떨까요?

GangWoon commented on ios/baseball/baseball/Presentor/UI/View/Play/MatchUpView.swift:
```diff
@@ -19,4 +26,14 @@ final class MatchUpView: UIView {
+    
+    func configurePitcherMatchUpView(_ pitcher: Pitcher) {
+        pitcherNameLabel.text = pitcher.name
+        numberOfPitchingLabel.text = "#\(pitcher.numberOfPitching)"
```
위 코맨트와 동일합니다.

GangWoon commented on ios/baseball/baseball/Presentor/UI/View/Score/PlayersHistoryCell.swift:
```diff
@@ -6,7 +6,25 @@
+//    }
+    
+    override func prepareForReuse() {
+        disposeBag = DisposeBag()
```
필요한 코드일까요?

GangWoon commented on ios/baseball/baseball/Presentor/UI/Controller/GameListViewController.swift:
```diff
@@ -0,0 +1,54 @@
+    }
+}
+
+//extension GameListViewController: UICollectionViewDataSource {
```
사용하지 않는 코드는 지워주세요.

Commit:
feat:RxSwift를 사용한 cell didSelect (#45)

GangWoon commented:
Outdated된 코드는 그냥 가볍게 읽고 넘어가시는 게 좋을 거 같습니다.
코드 빌드는 되시나요? 프로젝트를 다운받고 확인했을 때 빌드가 되지 않는 것고 노란색 줄이 많이 발견했습니다.
개인적으론 뷰를 너무 세분화하신 거 같습니다. ㅋㅋ 
추가적으로 PR 설명에는 해당 PR에 해당하는 부분만 작성하시는 게 좋습니다. 
읽는 사람에게 구현하지 않는 정보는 리뷰 집중력을 낮춥니다. 
전체적인 내용은 리드미에 따로 작성하시는게 어떨까요?

1. Codable을 채택한 대부분의 객체들을 private(set) var ...으로 변경하셨는데 이유를 들을 수 있을까요?
2. 메모리 누수는 발생하지 않으셨나요?
3. " ViewModel에서 Input - Output 구조를 시도해야 함. -> Test 유도, 재 사용성 매우 좋음)"은 어떤걸 말씀하시는지 구체적으로 표현해주실 수 있을까요?



> 이전 프로젝트에서 Combine을 사용하여, URLSession.dataTaskPublish를 사용 할 때와 현재 AlamoFire를 사용 할 때, 장점을 느 끼기 어렵습니다. AlamoFire를 옳게 사용하지 못 한 것인지 의아합니다.

RX...는 쉽게 생각해서 도구일 뿐입니다. 
개인적인 생각으론 외부 라이브러리를 사용하는 걸 꺼리지만 상황과 필요에 따라서 사용할 수 있으면 된다고 생각합니다.
다르게 말하자면 "알엑스 알라모파이어가 있었는데 정보조사가 미흡해서 아쉽다." 보다는 외부 라이브러리에 의존하지 않고 개발을 해보는 게 어떨까요?
기능을 구현하면서 다른 사람이 만들어 둔걸 사용하면 구현과정에서 겪는 어려움을 해결할 수 있는 기회를 놓친다는 게 개인적인 생각입니다.

제가 글을 잘쓰지 못해서 생각나는 것들을 정렬했습니다.
읽으시다가 이해가 안가시는 부분은 편하게 질문하셨으면 좋겠습니다.

godrm commented:
-----
추가적으로 Pods 폴더도 .gitignore에 추가해 주세요. `pod install` 작업으로 생기는 폴더는 저장소에 커밋하지 않는 것을 원칙으로 해주세요.
그렇지 않으면 패키지 버전 관리를 할 때마다 매번 충돌이 날 수도 있습니다.

Commit:
chore: pod gitignore 추가 (#none)

Commit:
Fix: 애매모호한 id 이름 수정 (#39)

Commit:
Refactor: 겹치는 UseCase Protocol 합치기 (#39)

Commit:
Fix: 사용하지 않는 코드 삭제 (#39)

Commit:
Refactor: 메서드 이름과 내부 내용 흐름이 일치하지 않는 부분 수정 (#39)

Commit:
Refactor: 메서드 이름과 내부 구현 맞지 않는 부분 수정 (#39)

Commit:
Refactor: 메서드 이름과 내부 구현 맞지 않는 부분 수정 (#39)

Commit:
Fix: 강제 언래핑 옵셔널 바인딩으로 수정 (#39)

dahun-lee-daji committed:
Merge pull request #47 from ghojeong/feat/ios/39/playTabArchitecture

[iOS] Network Layer 제외 피드백 반영

GangWoon commented on ios/baseball/baseball/Domain/UseCase/UseCasePort.swift:
```diff
@@ -0,0 +1,17 @@
+import RxSwift
+
+protocol UseCasePort {
+    func get<T: Codable>(path: APIPath, id: String?) -> Observable<T>
```
제네릭으로 잘만드셨는데 호출하는 부분에서 타입을 설정할순 없으셨나요?

dahun-lee-daji replied:
이 protocol과 관련해서는 추후 post 관련 내용을 추가 할 때 고려하여 진행하겠습니다!

GangWoon commented:
Presenter라고 잘 분류해두셨는데 엑션도 추상화 해보시는 건 어떨까요?
뷰에서 "나 특정 뷰를 보여줘"가 아니라 뷰에서 메세지를 보내고 뷰모델에서 메세지를 해석하는 방식으로요.
추가적으로 제가 단 코맨트가 정답이 아니에요. 질문을 받고 들었던 생각이나 접근 방법에 대해서 모르실 때 이모지보단 코맨트를 달아주셨으면 좋겠습니다. 

더 이상 고치실게 없다고 판단하셨다면 머지하셔도 괜찮을 거 같습니다.

dahun-lee-daji commented:
앗,, 코드가 빌드가 안되셨나요..? ㅜㅜ 백엔드 분들이 농담으로 로컬에선 잘됐는데..! 라고 하는 내용이 생각나네요,,, 제 생각에 아마 ScoreViewModel의 누락으로 인한게 아닌가 싶습니당!

저희는 위에 보여드린 체크박스 리스트를 이미 wiki, project로 관리중이고, 
전체적으로 진행을 어떻게 할 것인지 알려드리고 싶었는데 과도했던것 같습니다!ㅎㅎ
 
`Codable을 채택한 대부분의 객체들을 private(set) var ...으로 변경하셨는데 이유를 들을 수 있을까요?`
-> 
얻어온 데이터이니 변경할 수 없어야 하므로 (set) var이고,
상위 객체에서 하위 모델의 세부적인 데이터에 접근 할 수 없도록 하기 위해 private입니다!

`메모리 누수는 발생하지 않으셨나요?`
-> 혹시 네트워크 부분에서 우려를 표해 주신거라면, ViewModel에서 DisposeBag에 Dispose하므로 누수가 발생하지 않는다고 생각합니다!
저도 혹시나 싶으니 onCompleted를 추가해야겠네요!

`" ViewModel에서 Input - Output 구조를 시도해야 함. -> Test 유도, 재 사용성 매우 좋음)"은 어떤걸 말씀하시는지 구체적으로 표현해주실 수 있을까요?`
-> 아 이 내용은 저희가 본 예시 Project를 보고 막연히 생각한 내용입니다 ㅜ
ViewModel에서 struct 혹은 typealias Input, Output을 만들어 관리하며, 가독성과 유지보수 용이성을 얻으려 했던 내용입니다!

외부 라이브러리에 의존하지 않는 개발에 대한 내용에 공감합니다!
저도 항상 기본 프레임워크를 통해 class를 만들어 사용해왔었습니다. 
학습을 위해 외부 네트워크 라이브러리를 처음 사용해봐서 판단이 어려웠는데, 우문현답이네요 "필요에 따라서 사용해야 한다."
감사합니다!

dahun-lee-daji committed:
Refactor: FeedBack 반영 (#51)

1. enum APIPath 의 CustomStringConvertible 채택
2. 강제 언래핑 제거
3. 불필요한 코드 제거, memory leak 대비 onCompleted() 추가

ITzombietux committed:
Merge pull request #52 from ghojeong/Refactor/iOS/51/NetworkFeedback

Refactor: FeedBack 반영 (#51)

dahun-lee-daji commented:
추가적으로 수정 할 내용이 없는 것 같아서 merge 하겠습니다!

GangWoon commented:
알엑스에서 오퍼레이터를 사용하실 때 클로저를 강한 참조를하고 있어서 드렸던 질문입니다. 앱을 실행시키시면 오브젝트 그래프를 통해서 메모리 누수가 발상하고 있는지 확인하실수도 있습니다. 디스포즈백 내부가 어떻게 구현되어있는지 모르겠지만 약한 참조를 명시적으로 작성하시는게 혹시라도 발생할 수 있는 메모리 누수를 방지할 수 있는 방법입니다.
